File: 2020/14/Haskell/src/Part1.hs
module Part1 where

import           Data.Bits
import           Data.HashMap.Strict            ( HashMap )
import qualified Data.HashMap.Strict           as HM

type Memory = HashMap Int Int

applyMask :: [(Int, Char)] -> Int -> Int
applyMask []            value = value
applyMask ((a, v) : ms) value = applyMask ms value'
 where
  value' | v == '0'  = value `clearBit` a
         | otherwise = value `setBit` a

stepMemory :: String -> Memory -> [String] -> Memory
stepMemory _    memory []        = memory
stepMemory mask memory ([] : xs) = stepMemory mask memory xs
stepMemory mask memory (x : xs)
  | identifier == "mask" = stepMemory value memory xs
  | otherwise            = stepMemory mask memory' xs
 where
  [identifier, value] = filter (/= "=") . words $ x
  address             = read . reverse . drop 1 . reverse . drop 4 $ identifier
  mask' = filter ((`elem` "01") . snd) . zip [0 ..] . reverse $ mask
  value'              = applyMask mask' (read value)
  memory'             = HM.insert address value' memory

part1 :: [String] -> Int
part1 = sum . HM.elems . stepMemory [] HM.empty

File: 2020/14/Rust/src/part_1.rs
use crate::Instruction;

use std::collections::HashMap;

fn set_bit(num: u64, bit: u32) -> u64 {
    let bit = 2u64.pow(bit);
    num | bit
}

fn clear_bit(num: u64, bit: u32) -> u64 {
    let bit = 2u64.pow(bit);
    let complement = !bit;
    num & complement
}

fn apply_mask(mask: &str, mut value: u64) -> u64 {
    for (index, bit) in mask.chars().rev().enumerate() {
        if bit == '0' {
            value = clear_bit(value, index as u32);
        } else if bit == '1' {
            value = set_bit(value, index as u32);
        }
    }
    value
}

pub fn step_memory(instructions: &[Instruction]) -> u64 {
    let mut mask = String::new();
    let mut memory = HashMap::new();
    for instruction in instructions {
        let id = &instruction.id;
        let value = &instruction.value;

        if id == "mask" {
            mask = value.to_string();
        } else {
            let value = apply_mask(&mask, value.parse().unwrap());
            let id = id.parse::<u64>().unwrap();
            memory.insert(id, value);
        }
    }
    memory.values().sum()
}
